store/tikv/coprocessor.go

```diff
@@ -1189,3 +1225,151 @@ func (it copErrorResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
 func (it copErrorResponse) Close() error {
 	return nil
 }
+
+type rateLimitAction struct {
+	enabled        bool
+	fallbackAction memory.ActionOnExceed
+	once           sync.Once
+	// totalTokenNum indicates the total token at initial
+	totalTokenNum uint
+	cond          struct {
+		*sync.Cond
+		// exceed indicates whether have encountered OOM situation.
+		exceed bool
+		// existedTokenNum indicates the count of tokens which still exists
+		existedTokenNum uint
+		// isTokenDestroyed indicates whether there is one token has been isTokenDestroyed after Action
+		isTokenDestroyed bool
+	}
+}
+
+func newRateLimitAction(totalTokenNumber uint, cond *sync.Cond) *rateLimitAction {
+	return &rateLimitAction{
+		once:          sync.Once{},
+		totalTokenNum: totalTokenNumber,
+		cond: struct {
+			*sync.Cond
+			exceed           bool
+			existedTokenNum  uint
+			isTokenDestroyed bool
+		}{
+			Cond:            cond,
+			exceed:          false,
+			existedTokenNum: totalTokenNumber,
+		},
+	}
+}
+
+// Action implements ActionOnExceed.Action
+func (e *rateLimitAction) Action(t *memory.Tracker) {
+	if !e.enabled {
+		if e.fallbackAction != nil {
+			e.fallbackAction.Action(t)
+		}
+		return
+	}
+
+	e.once.Do(func() {
+		e.conditionLock()
+		defer e.conditionUnlock()
+		if e.cond.existedTokenNum < 2 {
+			logutil.BgLogger().Debug("rateLimitAction delegate to fallback action",
+				zap.Uint("totalTokenNum", e.totalTokenNum))
+			if e.fallbackAction != nil {
+				e.fallbackAction.Action(t)
+			}
+			return
+		}
+		logutil.BgLogger().Info("rateLimitAction exceed signal.",
+			zap.Int64("consumed", t.BytesConsumed()),
+			zap.Uint("existedToken", e.cond.existedTokenNum))
+		e.cond.isTokenDestroyed = false
+		e.cond.exceed = true
+	})
+}
+
+// SetLogHook implements ActionOnExceed.SetLogHook
+func (e *rateLimitAction) SetLogHook(hook func(uint64)) {
+
+}
+
+// SetFallback implements ActionOnExceed.SetFallback
+func (e *rateLimitAction) SetFallback(a memory.ActionOnExceed) {
+	e.fallbackAction = a
+}
+
+// broadcastIfNeeded will check whether the copWorkers is under suspended status.
+// If they are, `broadcastIfNeeded` would try to recover them if there are no more
+// copResponse remained in the channel.
+func (e *rateLimitAction) broadcastIfNeeded(needed bool) {
+	e.conditionLock()
+	defer e.conditionUnlock()
+	if e.cond.exceed && needed {
+		e.cond.exceed = false
+		e.cond.Broadcast()
+		e.once = sync.Once{}
```

Comment:
I think holding cond to reset once maybe race?  In Action(), the call stack is e.once.Do() -> e.cond.Lock()

Reply:
updated.